pkg/apis/sessions/session_state.go

```diff
@@ -29,7 +29,7 @@ type SessionState struct {
 	User              string   `msgpack:"u,omitempty"`
 	Groups            []string `msgpack:"g,omitempty"`
 	PreferredUsername string   `msgpack:"pu,omitempty"`
-
+	IntrospectClaims  string   `msgpack:"ic,omitempty"`
```

Comment:
In general we are trying to move away from adding new fields to this struct, however, in this case I think this is probably required as we have to make a separate API call to get this information.

Just to double check, we can't get this information from the profile URL or ID Token right? We want to add a generic claims map here to allow users to extract arbitrary data from the profile URL and ID Token and store that in the session, just want to check that won't be suitable

Reply:
Yup, this info is not available in the token or the user profile and the specific introspection request is needed to get access to this information.

In out case it contains membership of organisations and groups and such and can be quite large, hence it is not added to the token or profile.

Reply:
Are you concerned about the size of the session given this information could be quite large? If a user is using cookies for their session storage for example, will this bloat their cookies and cause issues? If we are increasing the session size, this definitely needs to be opt-in behaviour, which this should be already right?

Reply:
I'm also very curious about this given how much trouble we've had with session cookies in the past that are too large for IdPs that encode too much information in their tokens (looking at you Azure 😅 ).

Do you by chance have some data on how large this payload is in practice, and what it looks like? (we do lz4 compression so common fields/words will compress down well potentially).

Reply:
For background, this is what I remember from the bugs:

This issue isn't the 4kb cookie max size, we split it properly.  Where we get into trouble is the 8kb request header max size a lot of backend servers had.

Since the cookie isn't stripped from the request in going to the backend -- any session details you extract and add as headers (IDToken, Access Token, OIDC Introspective details) will then double what you already had, exacerbating the problem if they were large.

My question:

If this will be quite large, do we want to limit this to Redis session store only to avoid these very hard to debug bugs?

Reply:
I also didn't look at the code closely, but curious on this from the writeup:

> Introspection claims are stored as a base64 encoded string in the session state.

If that is still happening, you should change it:

Store your data raw in the SessionState. We then lz4 compress the MsgPack before encrypting it. THEN the raw encrypted bytes are base64 encoded to use in the cookie or redis session store.

If you Base64 ahead of time, you are removing our ability to getting effective LZ4 compression ratios and doing double encoding getting the 33% size penalty twice.

Reply:
If you need it base64 encoded for the upstream, I think you would wrap that operation here for the extraction:

```
case "introspect_claims":
		return []string{s.IntrospectClaims}
```

I'd need to defer to @JoelSpeed on that since the claim extractors are newer to me.

Reply:
Yeah I think you're correct nick, we can base64 encode it as we send it to the client as part of that big switch statement, assuming it must be base64 encoded for transport, if this isn't always true, do we need some switch to allow the end user to configure this?

Reply:
@NickMeves yeah will try to store `raw_data` in session state and encode while sending it to upstream

Reply:
> Are you concerned about the size of the session given this information could be quite large? If a user is using cookies for their session storage for example, will this bloat their cookies and cause issues? If we are increasing the session size, this definitely needs to be opt-in behaviour, which this should be already right?

Yes when we tried the max cookie size went to 12-15kb (3-4 cookies) were created.
As a side effect of this we encountered this issue https://github.com/oauth2-proxy/oauth2-proxy/issues/1209
Switched to redis and so far no issues.

Reply:
> For background, this is what I remember from the bugs:
> 
> This issue isn't the 4kb cookie max size, we split it properly. Where we get into trouble is the 8kb request header max size a lot of backend servers had.
> 
> Since the cookie isn't stripped from the request in going to the backend -- any session details you extract and add as headers (IDToken, Access Token, OIDC Introspective details) will then double what you already had, exacerbating the problem if they were large.
> 
> My question:
> 
> If this will be quite large, do we want to limit this to Redis session store only to avoid these very hard to debug bugs?

Yes, you are right.
We are facing this issue if the backend uses Nginx or some other web servers which returns 431 when the introspection value is too large.
Still trying to figure out a solution for this.

Reply:
So if you store this as raw bytes in the session and let us LZ4 & base64 encode just once, how is the sizing then? Ditching the double base64 will trim 25% off, LZ4 might give another 30-50% space savings?